File: tests/graph/test_factor_graph.py
"""Tests for FactorGraph.
"""
import random

import torch

from nsr.graph.factor_graph import LabelNode, FactorGraph


def test_unary_factor_node(unary_factor_node):
    """Test a factor node when it only touches two label nodes.
    """
    factor_node = unary_factor_node
    assert factor_node.score_table.shape == (9, )
    
    # Check score indexing
    factor_node.label_nodes[0].current_value = 3
    assert factor_node.update_score().item() == 3
    assert factor_node.current_score.item() == 3


def test_binary_factor_node(binary_factor_node):
    """Test a factor node when it touches two label nodes.
    """
    factor_node = binary_factor_node
    assert factor_node.score_table.shape == (9, 7)
    
    factor_node.label_nodes[0].current_value = 5
    factor_node.label_nodes[1].current_value = 2
    assert factor_node.update_score().item() == \
        factor_node.score_table[5, 2] == 37
    assert factor_node.score_table[3, 4] == 25


def test_graph_adjacency_list(five_node_graph):
    """Test if the graph is correctly built.
    """
    graph = five_node_graph
    assert graph.current_score is None
    
    assert len(graph.label_nodes) == 5
    for i, node in enumerate(graph.label_nodes):
        assert node.node_id == i
        assert node.label_space_size == 3
    assert len(graph.label_nodes[0].factor_nodes) == 2
    assert len(graph.label_nodes[1].factor_nodes) == 3
    assert len(graph.label_nodes[2].factor_nodes) == 3
    assert len(graph.label_nodes[3].factor_nodes) == 3
    assert len(graph.label_nodes[4].factor_nodes) == 3
    
    unary_nodes = graph.factor_nodes["unary"]
    assert len(unary_nodes) == 5
    for i, node in enumerate(unary_nodes):
        assert node.score_table_shape == [3]
        assert len(node.label_nodes) == 1
        assert node.label_nodes[0].node_id == i
        assert node in graph.label_nodes[i].factor_nodes
    assert graph.factor_label_indices["unary"] == [[0, 1, 2, 3, 4]]
    
    binary_nodes = graph.factor_nodes["binary"]
    assert len(binary_nodes) == 3
    assert [n.node_id for n in binary_nodes[0].label_nodes] == [1, 4]
    assert [n.node_id for n in binary_nodes[1].label_nodes] == [2, 3]
    assert [n.node_id for n in binary_nodes[2].label_nodes] == [3, 4]
    for factor_node in binary_nodes:
        assert factor_node.score_table_shape == [3, 3]
        for label_node in factor_node.label_nodes:
            assert factor_node in label_node.factor_nodes
    assert graph.factor_label_indices["binary"] == \
        [[1, 2, 3], [4, 3, 4]]

    tertiary_nodes = graph.factor_nodes["tertiary"]
    assert len(tertiary_nodes) == 1
    assert [n.node_id for n in tertiary_nodes[0].label_nodes] == [0, 1, 2]
    assert tertiary_nodes[0].score_table_shape == [3, 3, 3]
    for label_node in tertiary_nodes[0].label_nodes:
        assert tertiary_nodes[0] in label_node.factor_nodes
    assert graph.factor_label_indices["tertiary"] == [[0], [1], [2]]


def test_factor_graph_forward(five_node_graph):
    graph = five_node_graph
    input_states = torch.rand(5, 10)  # state_dim of the models is 10
    
    graph.forward(input_states)

    for node in graph.factor_nodes["unary"]:
        assert node.score_table.shape == (3, )
    for node in graph.factor_nodes["binary"]:
        assert node.score_table.shape == (3, 3)
    for node in graph.factor_nodes["tertiary"]:
        assert node.score_table.shape == (3, 3, 3)

    assert graph.current_score is None
    graph.random_init_labels()
    assert graph.current_score is not None


def test_factor_graph_compute_scores(five_node_graph_with_score):
    graph = five_node_graph_with_score
    graph.set_label_values([1, 0, 2, 1, 2])
    assert graph.current_score.item() == 31
    
    assert graph.compute_score_on_labels([0, 0, 0, 0, 0]).item() == 0
    assert graph.compute_score_on_labels([1, 0, 2, 1, 2]).item() == 31
    assert graph.compute_score_on_labels([2, 1, 0, 0, 1]).item() == 30
    assert graph.current_score.item() == 31  # Graph state is not changed


def test_init_with_oracle():
    graph = FactorGraph([10] * 100, dict(), dict())
    ground_truth = [random.randint(0, 9) for _ in range(100)]
    
    # Random init, mean 0.1
    graph.random_init_labels()
    assert sum([p == t for p, t in zip(graph.get_label_values(),
                                       ground_truth)]) < 30
    
    graph.random_init_labels(ground_truth, 1.0)
    assert sum([p == t for p, t in zip(graph.get_label_values(),
                                       ground_truth)]) == 100

    graph.random_init_labels(ground_truth, 0.5)
    assert sum([p == t for p, t in zip(graph.get_label_values(),
                                       ground_truth)]) > 30

File: cpp_tests/test_factor_graph.cpp
#include "factor_graph.hpp"
#include "cpp_tests.hpp"
#include "cpp_sampling.hpp"

#include <algorithm>


std::vector<float> test_unary_node() {
  std::vector<float> test_results;

  LabelNode* label = new LabelNode(3, 9);
  FactorNode* factor = new FactorNode({label});
  auto options = torch::TensorOptions().dtype(torch::kFloat32);
  torch::Tensor table = torch::arange(9, options);
  factor->set_score_table(table);
  
  factor->label_nodes[0]->current_value = 0;
  test_results.push_back(factor->update_score());  // Expect: 0.0

  factor->label_nodes[0]->current_value = 1;
  test_results.push_back(factor->update_score());  // Expect: 1.0

  factor->label_nodes[0]->current_value = 3;
  test_results.push_back(factor->update_score());  // Expect: 3.0

  std::vector<int> indices({7});
  test_results.push_back(factor->get_score(indices));  // Expect: 7.0
  factor->get_score(indices) = 101.7;
  test_results.push_back(factor->get_score(indices));  // Expect: 101.7

  delete label;
  delete factor;
  return test_results;
}


std::vector<float> test_binary_node() {
  std::vector<float> test_results;

  LabelNode* label_0 = new LabelNode(3, 9);
  LabelNode* label_1 = new LabelNode(24, 7);
  FactorNode* factor = new FactorNode({label_0, label_1});
  auto options = torch::TensorOptions().dtype(torch::kFloat32);
  torch::Tensor table = torch::arange(63, options);
  factor->set_score_table(table);

  factor->label_nodes[0]->current_value = 5;
  factor->label_nodes[1]->current_value = 2;
  test_results.push_back(factor->update_score());  // Expect 37.0
  
  std::vector<int> indices({5, 2});
  test_results.push_back(factor->get_score(indices));  // Expect 37.0
  indices = {3, 4};
  test_results.push_back(factor->get_score(indices));  // Expect 25.0
  indices = {8, 6};
  test_results.push_back(factor->get_score(indices));  // Expect 62.0

  delete label_0;
  delete label_1;
  delete factor;
  return test_results;
}


// Test fixture for the test cases
FactorGraph get_five_node_graph(bool set_scores) {
  std::vector<int> label_space_sizes({3, 3, 3, 3, 3});
  std::map<std::string, std::vector<std::vector<int>>> dependencies = {
    {"unary", {{0}, {1}, {2}, {3}, {4}}},
    {"binary", {{1, 4}, {2, 3}, {3, 4}}},
    {"tertiary", {{0, 1, 2}}}
  };
  FactorGraph graph = FactorGraph(label_space_sizes, dependencies);
  
  if (set_scores) {
    auto options = torch::TensorOptions().dtype(torch::kFloat32);
    torch::Tensor unary_table = torch::arange(3, options);
    torch::Tensor binary_table = torch::arange(9, options);
    torch::Tensor tertiary_table = torch::arange(27, options);
    for (FactorNode* node : graph.factor_nodes["unary"]) {
      node->set_score_table(unary_table);
    }
    for (FactorNode* node : graph.factor_nodes["binary"]) {
      node->set_score_table(binary_table);
    }
    for (FactorNode* node : graph.factor_nodes["tertiary"]) {
      node->set_score_table(tertiary_table);
    }
  }

  return graph;
}


std::vector<float> test_graph_adjacency_list() {
  std::vector<float> results;
  FactorGraph graph = get_five_node_graph();

  results.push_back(graph.label_nodes.size());  // Expect: 5.0
  for (int i = 0; i < 5; i++) {
    results.push_back(graph.label_nodes[i]->node_id);
    results.push_back(graph.label_nodes[i]->label_space_size);
    results.push_back(graph.label_nodes[i]->factor_nodes.size());
    // Expect: [2, 3, 3, 3, 3]
  }

  results.push_back(graph.factor_nodes.size());  // Expect: 3
  auto& unary_nodes = graph.factor_nodes["unary"];
  results.push_back(unary_nodes.size());  // Expect 5
  for (int i = 0; i < 5; i++) {
    FactorNode* node = unary_nodes[i];
    results.push_back(node->label_nodes.size());
    results.push_back(node->label_nodes[0]->node_id);
    auto& l_factors = graph.label_nodes[i]->factor_nodes;
    results.push_back(std::find(l_factors.begin(), l_factors.end(), node) 
      != l_factors.end());
  }

  auto& binary_nodes = graph.factor_nodes["binary"];
  results.push_back(binary_nodes.size());  // Expect 3
  for (FactorNode* node : binary_nodes) {
    for (LabelNode* l_node : node->label_nodes) {
      results.push_back(l_node->node_id);
      auto& l_factors = l_node->factor_nodes;
      results.push_back(std::find(l_factors.begin(), l_factors.end(), node) 
        != l_factors.end());
    }
  }

  auto& tertiary_nodes = graph.factor_nodes["tertiary"];  // Expect 1
  results.push_back(tertiary_nodes.size());
  for (LabelNode* l_node : tertiary_nodes[0]->label_nodes) {
    results.push_back(l_node->node_id);
    auto& l_factors = l_node->factor_nodes;
    results.push_back(std::find(l_factors.begin(), l_factors.end(), 
      tertiary_nodes[0]) != l_factors.end());
  }

  return results;
}


std::vector<float> test_factor_graph_compute_scores() {
  std::vector<float> results;
  FactorGraph graph = get_five_node_graph(true);
  
  std::vector<int> label_vals({1, 0, 2, 1, 2});
  results.push_back(graph.set_label_values(label_vals));
  results.push_back(graph.current_score);

  label_vals = {0, 0, 0, 0, 0};
  results.push_back(graph.compute_score_on_labels(label_vals));
  label_vals = {1, 0, 2, 1, 2};
  results.push_back(graph.compute_score_on_labels(label_vals));
  label_vals = {2, 1, 0, 0, 1};
  results.push_back(graph.compute_score_on_labels(label_vals));
  results.push_back(graph.current_score);

  return results;
}


// For this test case we wrap a unit of functionalities and assert in Python
std::vector<int> test_init_with_oracle(std::vector<int> ground_truth,
    float p, bool no_oracle) {
  // Create a dummy graph with Label nodes only
  std::vector<int> label_space_sizes;
  label_space_sizes.resize(100, 10);
  std::map<std::string, std::vector<std::vector<int>>> factor_dependencies;
  FactorGraph graph = FactorGraph(label_space_sizes, factor_dependencies);
  
  if (no_oracle) {
    graph.random_init_labels();
  }
  else {
    graph.random_init_labels(&ground_truth, p);
  }

  return graph.get_label_values();
}
